src/parsers/vc12DXMLLoad.cpp

```diff
@@ -0,0 +1,423 @@
+  if (!pItem->IsVoid())
+    globals.colour = vc12DXML_ToColour(pItem->AsString());
+}
+
+static void vc12DXML_Ammend_data_3d(udJSON const *pNode, std::vector<udDouble3> &pointList, vc12DXML_ProjectGlobals &globals)
+{
+  if (pNode->IsArray())
+  {
+    for (size_t i = 0; i < pNode->AsArray()->length; ++i)
+      vc12DXML_Ammend_data_3d(pNode->AsArray()->GetElement(i), pointList, globals);
+  }
+  else
+  {
+    char str[128] = {};
+    char *tokens[3] = {};
+    udDouble3 point = {};
+    udStrcpy(str, pNode->AsString());
+    if (vc12DXML_StrSplit(str, " ", tokens, 3) != 3)
+    {
+      LOG_WARNING("Malformed data_3d vertex item");
+    }
+    else
+    {
+      for (int p = 0; p < 3; ++p)
+      {
+        if (udStrcmp(tokens[p], "null") == 0)
+          point[p] = globals.nullValue;
+        else
+          point[p] = udStrAtof64(tokens[p]);
+      }
+      pointList.push_back(point);
+    }
+  }
+}
+
+vc12DXML_ProjectGlobals::vc12DXML_ProjectGlobals()
+  : nullValue(-999.0)                 // Default defined in the spec
+  , colour(vc12DXML_ToColour("red"))  // Default defined in the spec
+{
+
+}
+
+vc12DXML_Item::~vc12DXML_Item()
+{
+
+}
+
+vc12DXML_SuperString::vc12DXML_SuperString()
+  : m_pName(nullptr)
+  , m_isClosed(false)
+  , m_weight(1.0)
+  , m_colour(0xFFFF00FF)
+{
+
+}
+
+vc12DXML_SuperString::~vc12DXML_SuperString()
+{
+  udFree(m_pName);
+}
+
+udResult vc12DXML_SuperString::Build(udJSON const *pNode, vc12DXML_ProjectGlobals &globals)
+{
+  udResult result;
+  const udJSON *pItem = nullptr;
+  //bool has_z = false;
+  //double z = 0.0;
+
+  UD_ERROR_IF(pNode == nullptr, udR_InvalidParameter_);
+
+  // TODO Some values can and MUST be pulled out before others.
+  // <z> for example MUST be pulled out of the super_string block before reading any
+  // 3d/2d data, regardless of where <z> sits in the file in relation to the data.
+  // From the spec:
+  //   - There is exactly 0 or 1 <z> tag per string.
+  //   - Its position is irrelevant, it is global to the string
+  //pItem = &pNode->Get("z");
+  //UD_ERROR_IF(pItem->IsArray(), udR_InvalidConfiguration);
+  //
+  //if (!pItem->IsVoid())
+  //{
+  //  has_z = true;
+  //  z = pItem->AsDouble();
+  //}
+
+  SetGlobals(pNode, globals);
+
+  pItem = &pNode->Get("name");
+  if (!pItem->IsVoid())
+  {
+    udFree(m_pName);
+    m_pName = udStrdup(pItem->AsString());
+  }
+
+  pItem = &pNode->Get("closed");
+  if (!pItem->IsVoid())
+    m_isClosed = pItem->AsBool();
+
+  pItem = &pNode->Get("colour");
+  if (!pItem->IsVoid())
+    m_colour = vc12DXML_ToColour(pItem->AsString());
+  else
+    m_colour = globals.colour;
+
+  pItem = &pNode->Get("weight");
+  if (!pItem->IsVoid())
+    m_weight = pItem->AsDouble();
+
+  pItem = &pNode->Get("data_3d");
+  if (!pItem->IsVoid())
+    vc12DXML_Ammend_data_3d(&pItem->Get("p"), m_points, globals);
+
+  result = udR_Success;
+epilogue:
+  return result;
+}
+
+void vc12DXML_SuperString::Print(FILE *pOut, int indent) const
```

Comment:
Probably not. I used `FILE` as I was passing in `stdout`; not sure I could do that with a `udFile`. It is debug code. As far as keeping this code, it may be useful as more features are added, but am happy to remove.

Reply:
Yeah I don't think `udFile` could do that. The main reason to remove it is that it's exposed in the header as it's a member function, I'd prefer that `FILE` wasn't something that was just available, even if it's an internal header.